src/dbnode/storage/shard.go

```diff
@@ -1897,7 +1891,254 @@ func (s *dbShard) Flush(
 		multiErr = multiErr.Add(err)
 	}
 
-	return s.markFlushStateSuccessOrError(blockStart, version, multiErr.FinalError())
+	return s.markWarmFlushStateSuccessOrError(blockStart, multiErr.FinalError())
+}
+
+func (s *dbShard) ColdFlush(
+	flushPreparer persist.FlushPreparer,
+	resources coldFlushReuseableResources,
+) error {
+	// We don't flush data when the shard is still bootstrapping.
+	s.RLock()
+	if s.bootstrapState != Bootstrapped {
+		s.RUnlock()
+		return errShardNotBootstrappedToFlush
+	}
+	s.RUnlock()
+
+	dirtySeriesToWrite := resources.dirtySeriesToWrite
+	dirtySeries := resources.dirtySeries
+	idElementPool := resources.idElementPool
+	fsReader := resources.fsReader
+	// Reset resources.
+	for _, series := range dirtySeriesToWrite {
+		if series != nil {
+			series.Reset()
+		}
+		// Don't delete the empty list from the map so that other shards don't
+		// need to reinitialize the list for these blocks.
+	}
+	dirtySeries.Reset()
+	if fsReader.Status().Open {
+		if err := fsReader.Close(); err != nil {
+			return err
+		}
+	}
+
+	var multiErr xerrors.MultiError
+	// First, loop through all series to capture data on which blocks have dirty
+	// series and add them to the resources for further processing.
+	s.forEachShardEntry(func(entry *lookup.Entry) bool {
+		curr := entry.Series
+		seriesID := curr.ID()
+		blockStarts := curr.NeedsColdFlushBlockStarts()
+		blockStarts.ForEach(func(t xtime.UnixNano) {
+			seriesList := dirtySeriesToWrite[t]
+			if seriesList == nil {
+				seriesList = newIDList(idElementPool)
+				dirtySeriesToWrite[t] = seriesList
+			}
+			element := seriesList.PushBack(seriesID)
+
+			dirtySeries.Set(idAndBlockStart{blockStart: t, id: seriesID}, element)
+		})
+
+		return true
+	})
+
+	blockSize := s.namespace.Options().RetentionOptions().BlockSize()
+	srPool := s.opts.SegmentReaderPool()
+	multiIterPool := s.opts.MultiReaderIteratorPool()
+	encoderPool := s.opts.EncoderPool()
+	tmpCtx := context.NewContext()
+
+	// Loop through each block that we know has ColdWrites. Since each block
+	// has its own fileset, if we encounter an error while trying to persist
+	// a block, we continue to try persisting other blocks.
+BlockLoop:
```

Comment:
We should think about how to encapsulate this logic into a reusable way. Ideally, we could move most of this logic out into some kind of `FilesetMerger` or something.

The crux of the algorithm seems to be:

1. Loop through series on disk and check if they're in-memory. If in memory, merge, persist, then remove from in-memory lookup. If not in memory, just write them to disk.
2. Loop through series in memory lookup (which now has everything removed that was already merged / persisted to disk) and write that out to disk.

I think we'd need to discuss the interface a bit more but you can imagine something like:

```golang
type FsMergerOptions struct {
    Namespace ns.ID
    Shard uint32
    BlockStart time.Time
}

type FsMergeWith interface {
    Contains(seriesID ident.ID) (data checked.Bytes, bool)
    MarkMerged(seriesID ident.ID)
    ForEachRemaining(fn forEachRemainingFn) 
}

func (m *fsMerger) Merge(opts FsMergerOptions, with FsMergeWith) error {
   ...
}
```

I bet we could come up with a better interface, but I think you get the idea

Reply:
Then that thing could be tested fairly aggressively in isolation and this method would basically just boil down to calling the `FsMerger` with the right arguments.

Reply:
Thanks for the detailed suggestion. I've incorporated most of these ideas to the PR.